tests/lts_compatibility.py

```diff
@@ -0,0 +1,253 @@
+        # Note: Force a new node as primary by isolating LTS nodes
+        for node in lts_nodes:
+            node.suspend()
+
+        new_primary, _ = network.wait_for_new_primary(primary, nodes=new_nodes)
+        assert (
+            new_primary in new_nodes
+        ), "New node should have been elected as new primary"
+
+        for node in lts_nodes:
+            node.resume()
+
+        try:
+            network.wait_for_new_primary(new_primary)
+            assert False, "No new primary should be elected while"
+        except TimeoutError:
+            pass
+
+        # Retire one new node, so that at least one LTS node is required to reach consensus
+        other_new_nodes = [
+            node
+            for node in network.get_joined_nodes()
+            if (node is not new_primary and node in new_nodes)
+        ]
+        network.retire_node(new_primary, other_new_nodes[0])
+
+        issue_activity_on_live_service(network, args)
+
+        # Retire LTS nodes and code id
+        old_code_id = infra.utils.get_code_id(
+            args.enclave_type,
+            args.oe_binary,
+            args.package,
+            library_dir=lts_library_dir,
+        )
+        network.consortium.retire_code(primary, old_code_id)
+        for node in lts_nodes:
+            network.retire_node(new_primary, node)
+            node.stop()
+
+        issue_activity_on_live_service(network, args)
+
+
+def run_ledger_compatibility_since_first(args, use_snapshot):
+    """
+    Tests that a service from the very first LTS can be recovered
+    to the next LTS, and so forth, until the version of the local checkout.
+
+    The recovery process uses snapshot is `use_snapshot` is True. Otherwise, the
+    entire historical ledger is used.
+    """
+    repo = infra.gh_helper.Repository()
+    lts_releases = repo.get_lts_releases()
+
+    # TODO: Remove fakeness
+    lts_releases_fake = {}
+    lts_releases_fake["1.0"] = lts_releases["release/1.x"]
+    lts_releases_fake["2.0"] = lts_releases["release/1.x"]
+
+    # Add an empty entry to release to indicate local checkout
+    # Note: dicts are ordered from Python3.7
+    lts_releases_fake[None] = None
+
+    txs = app.LoggingTxs()
+    for idx, (_, lts_release) in enumerate(lts_releases_fake.items()):
+
+        if lts_release:
+            # version, lts_install_path = repo.install_release(lts_release)
+            # TODO: Fix
+            version = "1.0.0"
+            lts_install_path = "/data/git/CCF/build/ccf_install_1.0.0/opt/ccf"
+            binary_dir = os.path.join(lts_install_path, "bin")
+            library_dir = os.path.join(lts_install_path, "lib")
+            major_version = Version(version).release[0]
+            # Explictly test the logging app as it was packaged in the very first LTS
+            # TODO: Is this the right approach?
+            args.js_app_bundle = os.path.join(lts_install_path, "samples/logging/js")
```

Comment:
Do we want to test the latest version of the JS logging app here, or the one that was shipped with the very first LTS? I don't think we make claims about compatibility for JS apps yet, but a JS app built for `1.x` should just work (tm) on `N.x`?

Reply:
I agree that testing the one that shipped with the first LTS is the way to go, it's more likely to be run against a recent release than the latest sample against an LTS. The latest sample may legitimately make use of new features that aren't in the LTS.